nemo/collections/tts/modules/submodules.py

```
@@ -410,3 +418,244 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
+
+    @property
+    def input_types(self):
+        return {
+            "inputs": NeuralType(('B', 'D'), EncodedRepresentation()),
+            "token_id": NeuralType(('B'), Index(), optional=True),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "style_emb": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inputs):
+        bs = inputs.size(0)
+        query = inputs.unsqueeze(1)
+        key = F.tanh(self.tokens).unsqueeze(0).expand(bs, -1, -1)
+        value = F.tanh(self.tokens).unsqueeze(0).expand(bs, -1, -1)
+
+        style_emb, _ = self.mha(query=query, key=key, value=value,)
+        style_emb = style_emb.squeeze(1)
+        return style_emb
+
+
+class SpeakerEncoder(torch.nn.Module):
+    """
+    class SpeakerEncoder represents speakers representation. This module can combine GST (global style token)
+    based speaker embeddings, speaker embeddings from speaker verificaiton models and lookup table speaker
+    embeddings.
+    """
+
+    def __init__(self, gst_module=None, lookup_emb_projection_module=None):
+        """
+        Constructor.
+        Parameters:
+            - gst_module: Neural module to get GST based speaker embedding
+            - lookup_emb_projection_module: Neural module to project lookup speaker embedding
+        """
+        super(SpeakerEncoder, self).__init__()
+        self.gst_module = gst_module
+        self.lookup_emb_projection_module = lookup_emb_projection_module
+
+    def forward(self, spk_emb, reference_spec=None, reference_spec_lens=None):
+        x = spk_emb
+
+        # Project lookup speaker embedding
+        if self.lookup_emb_projection_module is not None:
+            out = self.lookup_emb_projection_module(spk_emb)
+            x = out if x is None else x + out
+
+        # Get GST based speaker embedding
+        if self.gst_module is not None and reference_spec is not None and reference_spec_lens is not None:
```

Should throw error if gst_module is not None and reference_spec is not provided.

This is also where you can include the GST errors/warnings that are in the FastPitchModel class right now.

---

Not sure is it good if we raise error inside the module? Since if we don't have that, the module will directly do. I write in FastPitchModel class to have an early check.

---

From an abstraction standpoint, nothing outside of the module should require any knowledge of what happens inside the model. Having error checks inside is also pretty foolproof.